### PR TITLE
[trainer] add tf32-mode control

### LINKED ISSUE #14450: [RFC] Amphere/tf32 defaults for transformers
## Background

It's possible to use the new TF32 format automatically when doing fp32 processing for a ~3x speed up, without doing any changes to the code, other than flipping the switch on. But the speed up may come at a cost of accuracy. You can see the differences between the formats in the following image:

tf32-Mantissa-chart-hi-res-FINAL-400x255

You can see that both TF32 and FP32 have the same dynamic range (the magnitude of numbers), but the former has a much lower precision, which depending on a situation may or may not impact the final outcome.

## Emerging Need

As Amphere hardware is emerging and TF32 automatic enabling seems to be going in the direction of being disabled by default (probably starting from pt-1.11?) as discussed here: https://github.com/pytorch/pytorch/issues/67384, we need to communicate to our users how to turn it on/off and what are the impacts on speed and accuracy might be.

Having it on could bring a ~3x speed improvement, and most likely according to the NVIDIA engineers the training quality shouldn't be impacted. But we don't have our first hand experiences yet to provide pragmatic recommendations.

## Available Guides

The on/off machinery is explained here: https://pytorch.org/docs/master/notes/cuda.html#tf32-on-ampere

The other crucial document is: https://pytorch.org/docs/master/notes/numerical_accuracy.html

TF32 Educational blog post: https://blogs.nvidia.com/blog/2020/05/14/tensorfloat-32-precision-format/

## Plan of action

So this issue is both an RFC and is also documenting the need to update: https://huggingface.co/transformers/performance.html

I trust that the commentary will emerge once you start experimenting with the new hardware.

@sgugger, @LysandreJik, @patil-suraj, @patrickvonplaten

### PR DESCRIPTION
**Pull request by stas00**
This PR adds tr32-mode control support for HF Trainer for Ampere cards. RFC: https://github.com/huggingface/transformers/issues/14450

pytorch had this mode on by default since pt-1.7, but are discussing to turn it off in the coming new release. https://github.com/pytorch/pytorch/issues/67384

Here is the proposed logic: 
1. by default HF Trainer will set it to Enabled,  This is marked as experimental should we discover that this is not a safe default down the road.
2. and `--tf32 0` will disable it. 
3. If the setup uses a wrong gpu or too low torch version - it will silently do nothing as it's irrelevant.

The PR adds:
- `is_torch_tf32_available` and `require_torch_tf32` helper utils
- adds basic test

Fixes: https://github.com/huggingface/transformers/issues/14450

@sgugger, @LysandreJik

### CODE REVIEW

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -548,6 +552,12 @@ class TrainingArguments:
         default=False,
         metadata={"help": "Whether to use full float16 evaluation instead of 32-bit"},
     )
+    tf32: bool = field(
+        default=True,
```
Comment (sgugger) [inline src/transformers/training_args.py:556]: The default should be whatever PyTorch has by default, so None here and the user can set it to True or False to force/unforce it.

I understand it's True for versions >= 1.7 and < 1.10 but False after?

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -802,6 +812,9 @@ def __post_init__(self):
                 "Mixed precision training with AMP or APEX (`--fp16` or `--bf16`) and half precision evaluation (`--fp16_full_eval` or `--bf16_full_eval`) can only be used on CUDA devices."
             )
 
+        if is_torch_available() and is_torch_tf32_available():
+            torch.backends.cuda.matmul.allow_tf32 = True if self.tf32 else False
```
Comment (sgugger) [inline src/transformers/training_args.py:816]: So here we should only change that boolean if the value set was not None. If the value is `True`, there should be an error if `is_torch_tf_32_available()` is `False` so the user is not surprised if they don't get what they want.

File: tests/test_trainer.py [hunk 3]
```diff
@@ -492,6 +493,15 @@ def test_mixed_bf16(self):
 
         # will add more specific tests once there are some bugs to fix
 
+    @require_torch_gpu
+    @require_torch_tf32
```
Comment (sgugger) [inline tests/test_trainer.py:497]: Out of curiosity, do we have a setup that has the right CUDA version an GPU capabilities?

**Review by sgugger — COMMENTED** [review]
As usual, let's operate on a strict no-breaking change rule. I understand PyTorch activates this feature by default when available (at least in some versions), so I would leave the default for `--tf32` to `None` and let PyTorch decide when `tf32=None`.

Then obviously `True` and `False` will force activate/deactivate the feature.

File: src/transformers/training_args.py (hunk 1 above)
Comment (stas00) [reply src/transformers/training_args.py:556]: ah, good idea! let the user decide!

File: tests/test_trainer.py (hunk 3 above)
Comment (stas00) [reply tests/test_trainer.py:497]: I have rtx-3090 if that's what you ask.

Running benchmarks now - will post those shortly.

File: src/transformers/training_args.py (hunk 2 above)
Comment (stas00) [reply src/transformers/training_args.py]: thanks for this great feedback, Sylvain. Please have another look.

File: src/transformers/training_args.py (hunk 1 above)
Comment (stas00) [reply src/transformers/training_args.py]: True for versions >= 1.7 and < 1.11 and probably False after - the nightly is still True as of today.

**Comment by stas00** [comment]
The benchmarks are terrible - going to ask for advice from the pytorch experts

File: docs/source/performance.md [hunk 4]
```diff
@@ -358,8 +358,13 @@ Like all cases with reduced precision this may or may not be satisfactory for yo
 
 If you're already using fp16 or bf16 mixed precision it may help with the throughput as well.
 
+You can enable this mode in the 🤗 Trainer with `--tf32`, or disable it with `--tf32 0`.
```
Comment (sgugger) [inline docs/source/performance.md:361]:
```suggestion
You can enable this mode in the 🤗 Trainer with `--tf32`, or disable it with `--tf32 0` or `--no_tf32`.
```

File: src/transformers/file_utils.py [hunk 5]
```diff
@@ -351,6 +351,24 @@ def is_torch_bf16_available():
         return False
 
 
+def is_torch_tf32_available():
+    if is_torch_available():
+        import torch
```
Comment (sgugger) [inline src/transformers/file_utils.py:356]: Tis is just a styling nit, but since we do a bunch of if not, return False afterward, could we remove that big test and begin with
```py
    if not is_torch_available():
        return False

    import torch
```
and then unindent all the other tests and remove the else part?

File: src/transformers/training_args.py [hunk 6]
```diff
@@ -227,6 +228,9 @@ class TrainingArguments:
         fp16_full_eval (:obj:`bool`, `optional`, defaults to :obj:`False`):
             Whether to use full float16 evaluation instead of 32-bit. This will be faster and save memory but can harm
             metric values.
+        tf32 (:obj:`bool`, `optional`, defaults to :obj:`None`):
```
Comment (sgugger) [inline src/transformers/training_args.py:231]:
```suggestion
        tf32 (:obj:`bool`, `optional`):
```
We never put the "defaults to `None`" as it's implied by the "optional".

File: src/transformers/training_args.py [hunk 7]
```diff
@@ -802,6 +812,17 @@ def __post_init__(self):
...
+        if is_torch_available() and self.tf32 is not None:
+            if self.tf32:
+                if is_torch_tf32_available():
+                    torch.backends.cuda.matmul.allow_tf32 = True
+                else:
+                    raise ValueError("--tf32 requires Ampere or a newer GPU arch, cuda>=11 and torch>=1.7")
+            else:
+                if is_torch_tf32_available():
+                    torch.backends.cuda.matmul.allow_tf32 = False
+                # nothing need to assert on else
```
Comment (sgugger) [inline src/transformers/training_args.py:824]: Perfect!

File: tests/test_trainer.py (hunk 3 above)
Comment (sgugger) [reply tests/test_trainer.py:497]: I was wondering for our testing machines on the automatic CI :-)

**Review by sgugger — APPROVED** [review]
Thanks a lot for the update, just left some styling nits but it's great!

File: tests/test_trainer.py (hunk 3 above)
Comment (stas00) [reply tests/test_trainer.py:497]: one day we will have those newer gpus.

File: src/transformers/file_utils.py (hunk 5 above)
Comment (stas00) [reply src/transformers/file_utils.py]: great suggestion, Sylvain. done.
